_rules/text-contrast-widget-nqzcj8.md

```diff
@@ -0,0 +1,443 @@
+---
+id: nqzcj8
+name: Text inside widget has minimum contrast
+rule_type: atomic
+description: |
+  This rule checks that, for text in widgets, the highest possible contrast of every text character with its background meets the minimal contrast requirement.
+accessibility_requirements:
+  wcag20:1.4.3: # Contrast (Minimum)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+  wcag20:1.4.6: # Contrast (Enhanced)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+input_aspects:
+  - Accessibility Tree
+  - DOM Tree
+  - CSS Styling
+  - Language
+acknowledgments:
+  authors:
+    - Jean-Yves Moyen
+---
+
+## Applicability
+
+This rule applies to any HTML element, character in a [text node][], and set of [widget pseudo-classes][] for which all the following are true:
```

Comment:
I don't think you should try to test elements, text nodes and pseudo-classes all in one rule. This seems way too complicated. You're having to do a lot of gymnastics to make this work, for example this list below, which should all be true, most of these things can only be applied to elements, or text nodes, but not to both. I think trying to do this all in one rule is going to make this more complicated than it needs to be.

Reply:
Ah, I see the confusion. The idea was not to apply the rule to any of these separately, but to any triplet composed of one of each. In order to check color contrast in a widget, you need to have both a text node and a "state" of the widget (and the widget itself make connection between them easy). Several conditions in the list are about how these 3 entities connect to each other.

Made that explicit. Should be better.